specs/_features/eip7732/beacon-chain.md

```diff
@@ -0,0 +1,716 @@
+    """
+    Return the set of attesting indices corresponding to ``payload_attestation``.
+    """
+    ptc = get_ptc(state, slot)
+    return set(index for i, index in enumerate(ptc) if payload_attestation.aggregation_bits[i])
+```
+
+#### `get_indexed_payload_attestation`
+
+```python
+def get_indexed_payload_attestation(state: BeaconState, slot: Slot, 
+                                    payload_attestation: PayloadAttestation) -> IndexedPayloadAttestation:
+    """
+    Return the indexed payload attestation corresponding to ``payload_attestation``.
+    """
+    attesting_indices = get_payload_attesting_indices(state, slot, payload_attestation)
+
+    return IndexedPayloadAttestation(
+        attesting_indices=sorted(attesting_indices),
+        data=payload_attestation.data,
+        signature=payload_attestation.signature,
+    )
+```
+
+## Beacon chain state transition function
+
+*Note*: state transition is fundamentally modified in EIP-7732. The full state transition is broken in two parts, first importing a signed block and then importing an execution payload.
+
+The post-state corresponding to a pre-state `state` and a signed beacon block `signed_block` is defined as `state_transition(state, signed_block)`. State transitions that trigger an unhandled exception (e.g. a failed `assert` or an out-of-range list access) are considered invalid. State transitions that cause a `uint64` overflow or underflow are also considered invalid.
+
+The post-state corresponding to a pre-state `state` and a signed execution payload envelope `signed_envelope` is defined as `process_execution_payload(state, signed_envelope)`. State transitions that trigger an unhandled exception (e.g. a failed `assert` or an out-of-range list access) are considered invalid. State transitions that cause an `uint64` overflow or underflow are also considered invalid.
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    process_withdrawals(state)  # [Modified in EIP-7732]
+    process_execution_payload_header(state, block)  # [Modified in EIP-7732, removed process_execution_payload]
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)
+    process_operations(state, block.body)  # [Modified in EIP-7732]
+    process_sync_aggregate(state, block.body.sync_aggregate)
+```
+
+#### Withdrawals
+
+##### Modified `process_withdrawals`
+
+**Note:** This is modified to take only the `state` as parameter. Withdrawals are deterministic given the beacon state, any execution payload that has the corresponding block as parent beacon block is required to honor these withdrawals in the execution layer. This function must be called before `process_execution_payload_header` as this latter function affects validator balances. 
+
+```python
+def process_withdrawals(state: BeaconState) -> None:
+    # return early if the parent block was empty
```

Comment:
This has an interesting scenario which may warrant a test:

1. Balances are decreased in the CL but are only increased in the EL at a later stage once the payload is revealed.
2. Assume that revealing fails for a while for some reason, but the CL still continues. Meaning the balances have been decreased in the CL but withdrawals are still pending.
3. Assume this goes on for like 20 minutes and the chain finalizes.

Someone doing checkpoint sync from the finalized checkpoint will now no longer be able to checkpoint sync, because they do not know the actual `withdrawals` from `get_expected_withdrawals` as they have already been processed in the CL. Because they only have data from finalized checkpoint onward, they are now unable to send the expected withdrawals to an EL and will not be able to propose any blocks until someone else who still has the data recovers the situation and manages to properly reveal a payload.